control-plane-operator/controllers/hostedcontrolplane/kas/auth.go

```diff
@@ -48,50 +53,255 @@ func generateAuthConfig(spec *configv1.AuthenticationSpec, ctx context.Context,
 		}
-		jwt := JWTAuthenticator{
-			Issuer: Issuer{
-				URL:                  provider.Issuer.URL,
-				CertificateAuthority: caData,
-			},
+		config.JWT = append(config.JWT, jwt)
+	}
+	return config, nil
+}
+
+func generateJWTForProvider(ctx context.Context, provider configv1.OIDCProvider, client crclient.Client, namespace string) (JWTAuthenticator, error) {
+	out := JWTAuthenticator{}
+
+	issuer, err := generateIssuer(ctx, provider.Issuer, client, namespace)
+	if err != nil {
+		return out, fmt.Errorf("generating issuer: %v", err)
+	}
+
+	claimMappings, err := generateClaimMappings(provider.ClaimMappings, issuer.URL)
+	if err != nil {
+		return out, fmt.Errorf("generating claim mappings: %v", err)
+	}
+
+	claimValidationRules, err := generateClaimValidationRules(provider.ClaimValidationRules...)
+	if err != nil {
+		return out, fmt.Errorf("generating claim validation rules: %v", err)
+	}
+
+	out.Issuer = issuer
+	out.ClaimMappings = claimMappings
+	out.ClaimValidationRules = claimValidationRules
+
+	return out, nil
+}
+
+func generateIssuer(ctx context.Context, issuer configv1.TokenIssuer, client crclient.Client, namespace string) (Issuer, error) {
+	out := Issuer{}
+
+	out.URL = issuer.URL
+	out.AudienceMatchPolicy = AudienceMatchPolicyMatchAny
+
+	for _, audience := range issuer.Audiences {
+		out.Audiences = append(out.Audiences, string(audience))
+	}
+
+	if len(issuer.CertificateAuthority.Name) > 0 {
+		ca, err := getCertificateAuthorityFromConfigMap(ctx, client, issuer.CertificateAuthority.Name, namespace)
+		if err != nil {
+			return out, fmt.Errorf("getting certificate authority for issuer: %v", err)
 		}
-		audience := []string{}
-		for _, a := range provider.Issuer.Audiences {
-			audience = append(audience, string(a))
+		out.CertificateAuthority = ca
+	}
+
+	return out, nil
+}
+
+func getCertificateAuthorityFromConfigMap(ctx context.Context, client crclient.Client, caName, namespace string) (string, error) {
+	ca := &corev1.ConfigMap{}
+	if err := client.Get(ctx, crclient.ObjectKey{Name: caName, Namespace: namespace}, ca); err != nil {
+		return "", fmt.Errorf("failed to get issuer certificate authority configmap: %w", err)
+	}
+
+	caData, ok := ca.Data[certificateAuthorityConfigMapKey]
+	if !ok {
+		return "", fmt.Errorf("issuer certificate authority configmap does not contain key %q", certificateAuthorityConfigMapKey)
+	}
+
+	return caData, nil
+}
+
+func generateClaimMappings(claimMappings configv1.TokenClaimMappings, issuerURL string) (ClaimMappings, error) {
+	out := ClaimMappings{}
+
+	username, err := generateUsernameClaimMapping(claimMappings.Username, issuerURL)
+	if err != nil {
+		return out, fmt.Errorf("generating username claim mapping: %v", err)
+	}
+
+	groups := generateGroupsClaimMapping(claimMappings.Groups)
+
+	out.Username = username
+	out.Groups = groups
+
+	if featuregate.Gates.Enabled(featuregate.ExternalOIDCWithUIDAndExtraClaimMappings) {
```

Comment:
TODO: pending https://github.com/openshift/hypershift/pull/5976 merging, will need to update how the feature gates are being used here